klippy/gcode.py

```diff
@@ -349,6 +360,24 @@ def _process_data(self, eventtime):
         if self.fd_handle is None:
             self.fd_handle = self.reactor.register_fd(self.fd,
                                                       self._process_data)
+    def _handle_remote_help(self, web_request):
+        web_request.send(dict(self.gcode_help))
+    def _handle_remote_restart(self, web_request):
+        path = web_request.get_path()
+        if path == "gcode/restart":
+            web_request.put('script', 'restart')
+        elif path == "gcode/firmware_restart":
+            web_request.put('script', 'firmware_restart')
+        self.run_script_from_remote(web_request)
```

Comment:
Could you elaborate on this support?  The `put()` method looks a bit odd.  Also, is there an advantage to having an endpoint that just runs a script instead of having the caller just issue the script directly?  (That is, why issue `gcode/restart` vs `gcode/script('RESTART')`?)

Just thinking out loud, I'd think we'd want to use webhook endpoints if: the endpoint produces data (as parsing gcode responses would be a pain), or if the command needs to run immediately without waiting for previous gcode commands to complete.

Reply:
The put() method adds an argument to the web_request so `run_script_from_remote` would work correctly.  This was for done for exeption handling, but I don't believe that will be necessary with your suggested changes to the exception handling in webhooks.py, it would be fine to simply call run_script().

With regard to having an endpoint that simply runs a script, the primary purpose is to provide common APIs for web clients.  They could use `/printer/gcode/script`, but exposing certain methods via the API increases their visibility.  If you would prefer to remove these endpoints I don't have a problem with it, functionally they are equivalent.

Reply:
Okay.  I'm fine either way wrt adding the additional endpoints.  I'd recommend against the put() method though, as I think it may be a little confusing to have the code alter the request state.  Perhaps have a helper like `gcode.run_script_from_webhook(my_script, webhook)`.  So that the script is passed explicitly.

Reply:
Yes, I am removing the put() method.  Since the webhooks module will handle `homing.CommandError` I should be able to call gcode.run_script() without the need for an additional helper.